File: web/src/services/api/evolution.ts
import { api } from './client'

export type EvolutionEventResponse = {
  id: string
  faq_id: string
  version_id: string | null
  event_type: string
  description: string | null
  triggered_by: string | null
  created_at: string
}

export type EvolutionEventListResponse = {
  items: EvolutionEventResponse[]
  total: number
}

export type EvolutionTimelineEntry = {
  version_id: string
  version_number: number
  title: string
  change_summary: string | null
  created_by: string
  created_at: string
  is_current: boolean
  diff_summary: string | null
}

export type FaqTimelineResponse = {
  faq_id: string
  faq_title: string
  current_version: number
  timeline: EvolutionTimelineEntry[]
  events: EvolutionEventResponse[]
}

export type VersionDiffResponse = {
  faq_id: string
  from_version: number
  to_version: number
  diff: { op: string; before: string[]; after: string[] }[]
  additions: number
  deletions: number
}

export type SynthesisResponse = {
  discussion_id: string
  candidate_id: string | null
  title: string
  confidence_score: number
  source_reply_ids: string[]
  used_fallback: boolean
}

export const evolutionApi = {
  listEvents: (limit = 50) =>
    api.get<EvolutionEventListResponse>('/api/v1/evolution/events', { limit: String(limit) }),

  getTimeline: (faqId: string) =>
    api.get<FaqTimelineResponse>(`/api/v1/evolution/timeline/${faqId}`),

  getDiff: (faqId: string, fromV: number, toV: number) =>
    api.get<VersionDiffResponse>(`/api/v1/evolution/diff/${faqId}/${fromV}/${toV}`),

  synthesize: (discussionId: string, force = false) =>
    api.post<SynthesisResponse>(
      `/api/v1/discussions/${discussionId}/synthesize`,
      { force },
    ),

  rollback: (faqId: string, targetVersionId: string) =>
    api.post<{ faq_id: string; current_version: number; title: string }>(
      `/api/v1/faqs/${faqId}/rollback`,
      { target_version_id: targetVersionId },
    ),
}

File: backend/app/schemas/evolution.py
"""Pydantic schemas for the Knowledge Evolution API."""
from datetime import datetime
from pydantic import BaseModel, Field

from app.schemas.base import BaseSchema


class EvolutionEventResponse(BaseSchema):
    id: str
    faq_id: str
    version_id: str | None
    event_type: str
    description: str | None
    triggered_by: str | None
    created_at: datetime


class EvolutionEventListResponse(BaseModel):
    items: list[EvolutionEventResponse]
    total: int


class EvolutionTimelineEntry(BaseModel):
    version_id: str
    version_number: int
    title: str
    change_summary: str | None
    created_by: str
    created_at: datetime
    is_current: bool
    diff_summary: str | None = None


class FaqTimelineResponse(BaseModel):
    faq_id: str
    faq_title: str
    current_version: int
    timeline: list[EvolutionTimelineEntry]
    events: list[EvolutionEventResponse]


class VersionDiffResponse(BaseModel):
    faq_id: str
    from_version: int
    to_version: int
    diff: list[dict] = Field(default_factory=list)
    additions: int = 0
    deletions: int = 0


class RollbackRequest(BaseModel):
    target_version_id: str


class SynthesisRequest(BaseModel):
    force: bool = False


class SynthesisResponse(BaseModel):
    discussion_id: str
    candidate_id: str | None
    title: str
    confidence_score: float
    source_reply_ids: list[str]
    used_fallback: bool
